tpm2/policy_v0.go

```diff
@@ -68,10 +69,184 @@ func computeV0PinNVIndexPostInitAuthPolicies(alg tpm2.HashAlgorithmId, updateKey
+// a slice of nodes, with each node specifying an offset to its parent node.
+type policyOrData_v0 []*policyOrDataNode_v0
+
+type policyOrDataResolver_v0 struct {
+	src         []*policyOrDataNode_v0
+	all         []*policyOrNode
+	numResolved int
+	depth       int
+}
+
+func (r *policyOrDataResolver_v0) resolveOne(n uint32) (*policyOrNode, error) {
+	// newPolicyOrTree limits the depth by limiting the
+	// number of digests. Limit the depth when constructing
+	// a tree from its serialized format too.
+	r.depth += 1
+	defer func() { r.depth -= 1 }()
+	if r.depth > policyOrMaxDepth {
+		return nil, errors.New("too deep")
+	}
+
+	if int(n) >= len(r.all) {
+		// A node is indexing its parent from outside of the tree.
+		return nil, fmt.Errorf("index %d out of range", n)
+	}
+
+	node := r.all[n]
+	if node == nil {
+		// This is a node we haven't resolved yet.
+		// Create the resolved node and copy the digests across.
+		node = new(policyOrNode)
+		node.digests = make(tpm2.DigestList, len(r.src[n].Digests))
+		copy(node.digests, r.src[n].Digests)
+
+		// Bounds check the offset to the parent node.
+		if r.src[n].Next > policyOrMaxDigests/8 {
+			return nil, fmt.Errorf("next value too large (%d)", r.src[n].Next)
+		}
+
+		if r.src[n].Next != 0 {
+			// This is not the root node. Resolve the parent node.
+			parent, err := r.resolveOne(n + r.src[n].Next)
+			if err != nil {
+				return nil, err
+			}
+			node.parent = parent
+		}
+
+		// Save the resolved node so we can return the same one for
+		// other children.
+		r.numResolved += 1
+		r.all[n] = node
+	}
+	return node, nil
+}
+
+// resolve converts the flattened tree in to an actual tree, suitable for
+// executing assertions on.
+func (t policyOrData_v0) resolve() (out *policyOrTree, err error) {
+	if len(t) == 0 {
+		return nil, errors.New("no nodes")
+	}
+
+	// The offset to the first non-leaf node tells us how many leaf
+	// nodes we have.
+	numLeafNodes := t[0].Next
+	if numLeafNodes == 0 {
+		// This is a tree containing a single node.
+		numLeafNodes = 1
+	}
+
+	if numLeafNodes > policyOrMaxDigests/8 {
+		return nil, fmt.Errorf("too many leaf nodes (%d)", numLeafNodes)
+	}
+
+	resolver := &policyOrDataResolver_v0{
+		src: t,
+		all: make([]*policyOrNode, len(t))}
+
+	out = new(policyOrTree)
+
+	// Resolve each leaf node.
+	for i := uint32(0); i < numLeafNodes; i++ {
+		leaf, err := resolver.resolveOne(i)
+		if err != nil {
+			return nil, err
+		}
+		out.leafNodes = append(out.leafNodes, leaf)
+	}
+
+	if resolver.numResolved < len(t) {
+		// The data contains nodes that aren't reachable, so consider it to
+		// be invalid.
+		return nil, errors.New("unreachable nodes")
+	}
+
+	return out, nil
+}
+
+// newPolicyOrDataV0 creates a new flattened tree from the supplied digests
+// for creating a policy that can be satisified by multiple conditions. It also
+// extends the supplied trial policy.
+//
+// It works by turning the supplied list of digests (each corresponding to some
```

Comment:
this part of the comment seems to need changing to refer to the tree and not digests